File: Day2/main.hs
import Data.Char
import Data.Function
import System.IO

normalise c =
  let i = ord c
   in if i >= ord 'X' then i - ord 'X' else i - ord 'A'

playRound (a, b)
  | beat a == b = 6 + b + 1
  | a == b = 3 + b + 1
  | otherwise = b + 1

beat a = (a + 1) `mod` 3

lose 0 = 2
lose a = a - 1

playRound' (a, c) =
  let b = case c of
        0 -> lose a
        1 -> a
        2 -> beat a
   in playRound (a, b)

part1 contents =
  lines contents
    & map (\line -> playRound (normalise $ head line, normalise $ line !! 2))
    & sum
    & show

part2 contents =
  lines contents
    & map (\line -> playRound' (normalise $ head line, normalise $ line !! 2))
    & sum
    & show

main = do
  small_contents <- readFile "in.small"
  contents <- readFile "in"

  putStrLn $ part1 small_contents
  putStrLn $ part1 contents

  putStrLn $ part2 small_contents
  putStrLn $ part2 contents

File: Day2/main.exs
defmodule Day do
  def normalise(hand) do
    case hand do
      "X" -> 0
      "Y" -> 1
      "Z" -> 2
      "A" -> 0
      "B" -> 1
      "C" -> 2
    end
  end

  def beat(hand) do
    Integer.mod(hand + 1, 3)
  end

  def lose(hand) do
    case hand do
      0 -> 2
      x -> x - 1
    end
  end

  def part1(input) do
    input
    |> Stream.map(fn line ->
      a = normalise(String.at(line, 0))
      b = normalise(String.at(line, 2))

      b + 1 +
        cond do
          beat(a) === b -> 6
          a === b -> 3
          true -> 0
        end
    end)
    |> Enum.sum()
  end

  def part2(input) do
    input
    |> Stream.map(fn line ->
      a = normalise(String.at(line, 0))
      c = normalise(String.at(line, 2))

      b =
        case c do
          0 -> lose(a)
          1 -> a
          2 -> beat(a)
        end

      b + 1 +
        cond do
          beat(a) === b -> 6
          a === b -> 3
          true -> 0
        end
    end)
    |> Enum.sum()
  end
end

small_input = File.stream!("in.small")
input = File.stream!("in")

IO.puts(Day.part1(small_input))
IO.puts(Day.part1(input))

IO.puts(Day.part2(small_input))
IO.puts(Day.part2(input))
